scripts/check-public-api.ts

```diff
@@ -0,0 +1,359 @@
+    const exported = captureGroupsFromGlobalRegex(regex, normalized).map(
+      element => ({ name: element, type: objectType })
+    );
+    return [...allObjects, ...exported];
+  }, []);
+}
+
+/**
+ * Parse a barrel file for the exported objects.
+ * It selects all string in \{\} e.g. export \{a,b,c\} from './xyz' will result in [a,b,c].
+ * @param fileContent - Content of the index file to be parsed.
+ * @param regex - Regular expression used for matching exports.
+ * @returns List of objects exported by the given index file.
+ */
+export function parseBarrelFile(fileContent: string, regex: RegExp): string[] {
+  const normalized = fileContent.replace(/\s+/g, '');
+  const groups = captureGroupsFromGlobalRegex(regex, normalized);
+
+  return flatten(groups.map(group => group.split(',')));
+}
+
+function checkInternalReExports(fileContent: string, filePath: string): void {
+  const internalReExports = parseBarrelFile(
+    fileContent,
+    /\{([\w,]+)\}from'.*\/internal'/g
+  );
+  if (internalReExports.length) {
+    throw new Error(
+      `Re-exporting internal modules is not allowed. ${internalReExports
+        .map(reExport => `'${reExport}'`)
+        .join(', ')} exported in '${filePath}'.`
+    );
+  }
+}
+
+export async function parseIndexFile(filePath: string): Promise<string[]> {
+  const cwd = dirname(filePath);
+  const fileContent = await readFile(filePath, 'utf-8');
+  checkInternalReExports(fileContent, filePath);
+  const localExports = parseBarrelFile(fileContent, regexExportedIndex);
+  const starFiles = captureGroupsFromGlobalRegex(
+    /export \* from '([\w\/.-]+)'/g,
+    fileContent
+  );
+  const starFileExports = await Promise.all(
+    starFiles.map(async relativeFilePath => {
+      const fullPath = resolve(cwd, relativeFilePath);
+      const tsPath = fullPath.replace(/\.js$/, '.ts');
+      return parseIndexFile(tsPath);
+    })
+  );
+  return [...localExports, ...starFileExports.flat()];
+}
+
+function captureGroupsFromGlobalRegex(regex: RegExp, str: string): string[] {
+  const groups = Array.from(str.matchAll(regex));
+  return groups.map(group => group[1]);
+}
+
+export async function checkBarrelRecursive(cwd: string): Promise<void> {
+  (await readdir(cwd, { withFileTypes: true }))
+    .filter(dirent => dirent.isDirectory())
+    .forEach(async subDir => {
+      if (['__snapshot__', 'spec'].includes(subDir.name)) {
+        await checkBarrelRecursive(join(cwd, subDir.name));
+      }
+    });
+  await exportAllInBarrel(
+    cwd,
+    parse(cwd).name === 'src' ? 'internal.ts' : 'index.ts'
+  );
+}
+
+export async function exportAllInBarrel(
+  cwd: string,
+  barrelFileName: string
+): Promise<void> {
+  const barrelFilePath = join(cwd, barrelFileName);
+  if (existsSync(barrelFilePath) && (await lstat(barrelFilePath)).isFile()) {
+    const dirContents = (
+      await glob('*', {
+        ignore: [
+          '**/*.spec.ts',
+          '**/*.test.ts',
```

Comment:
I see you added .test.ts, however, I think you can also remove .spec.ts (since we do not use it)